sql/expression/function/time.go

```diff
@@ -316,6 +328,200 @@ func datePartFunc(fn func(time.Time) int) func(interface{}) interface{} {
+func (d *YearWeek) IsNullable() bool {
+	return d.date.IsNullable()
+}
+
+// TransformUp implements the Expression interface.
+func (d *YearWeek) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	date, err := d.date.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+
+	mode, err := d.mode.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+
+	yw, err := NewYearWeek(date, mode)
+	if err != nil {
+		return nil, err
+	}
+	return f(yw)
+}
+
+// Following solution of YearWeek was taken from tidb: https://github.com/pingcap/tidb/blob/master/types/mytime.go
+type weekBehaviour int64
+
+const (
+	// weekBehaviourMondayFirst set Monday as first day of week; otherwise Sunday is first day of week
+	weekBehaviourMondayFirst weekBehaviour = 1 << iota
+	// If set, Week is in range 1-53, otherwise Week is in range 0-53.
+	// Note that this flag is only relevant if WEEK_JANUARY is not set.
+	weekBehaviourYear
+	// If not set, Weeks are numbered according to ISO 8601:1988.
+	// If set, the week that contains the first 'first-day-of-week' is week 1.
+	weekBehaviourFirstWeekday
+)
+
+func (v weekBehaviour) test(flag weekBehaviour) bool {
+	return (v & flag) != 0
+}
+
+func weekMode(mode int64) weekBehaviour {
+	weekFormat := weekBehaviour(mode & 7)
+	if (weekFormat & weekBehaviourMondayFirst) == 0 {
+		weekFormat ^= weekBehaviourFirstWeekday
+	}
+	return weekFormat
+}
+
+// calcWeekday calculates weekday from daynr, returns 0 for Monday, 1 for Tuesday ...
+func calcWeekday(daynr int32, sundayFirstDayOfWeek bool) int32 {
+	daynr += 5
+	if sundayFirstDayOfWeek {
+		daynr++
+	}
+	return daynr % 7
+}
+
+// calcWeek calculates week and year for the time.
+func calcWeek(yyyy, mm, dd int32, wb weekBehaviour) (int32, int32) {
+	daynr := calcDaynr(yyyy, mm, dd)
+	firstDaynr := calcDaynr(yyyy, 1, 1)
+	mondayFirst := wb.test(weekBehaviourMondayFirst)
+	weekYear := wb.test(weekBehaviourYear)
+	firstWeekday := wb.test(weekBehaviourFirstWeekday)
+	weekday := calcWeekday(firstDaynr, !mondayFirst)
+
+	week, days := int32(0), int32(0)
+	if mm == 1 && dd <= 7-weekday {
+		if !weekYear &&
+			((firstWeekday && weekday != 0) || (!firstWeekday && weekday >= 4)) {
+			return yyyy, week
+		}
+		weekYear = true
+		yyyy--
+		days = calcDaysInYear(yyyy)
+		firstDaynr -= days
+		weekday = (weekday + 53*7 - days) % 7
+	}
+
+	if (firstWeekday && weekday != 0) ||
+		(!firstWeekday && weekday >= 4) {
+		days = daynr - (firstDaynr + 7 - weekday)
+	} else {
+		days = daynr - (firstDaynr - weekday)
+	}
+
+	if weekYear && days >= 52*7 {
+		weekday = (weekday + calcDaysInYear(yyyy)) % 7
+		if (!firstWeekday && weekday < 4) ||
+			(firstWeekday && weekday == 0) {
+			yyyy++
+			week = 1
+			return yyyy, week
+		}
+	}
+	week = days/7 + 1
+	return yyyy, week
+}
+
+// calcDaysInYear calculates days in one year, it works with 0 <= yyyy <= 99.
+func calcDaysInYear(yyyy int32) int32 {
+	if (yyyy&3) == 0 && (yyyy%100 != 0 || (yyyy%400 == 0 && (yyyy != 0))) {
```

Comment:
It looks like a calculation of isLeapYear is it?

Reply:
yes, it's